src/common/logging.rs

```diff
@@ -0,0 +1,194 @@
+};
+
+use self::stage2_config::Stage2Config;
+
+use super::{
+    defs::{BALENA_PART_MP, OLD_ROOT_MP},
+    stage2_config,
+};
+
+/// Utility function to create a directory and all child directories
+fn create_dir_if_not_exist(path: &str) -> Result<()> {
+    let dir_path = Path::new(path);
+    if !dir_path.is_dir() {
+        println!("Directory does not exist. Creating: {}", dir_path.display());
+        fs::create_dir_all(dir_path).upstream_with_context(
+            format!("Failed to create directory {}", dir_path.display()).as_str(),
+        )?;
+        println!("Directory created successfully.");
+    } else {
+        println!("Directory already exists: {}", dir_path.display());
+    }
+
+    Ok(())
+}
+
+/// Helper function to copy a source file to a directory
+/// it keeps the same file name
+pub fn copy_file_to_destination_dir(source_file_path: &str, dest_dir_path: &str) -> Result<()> {
+    info!(
+        "copy_file_to_destination_dir! Copying {} from tmpfs to {}",
+        source_file_path, dest_dir_path
+    );
+
+    let source_file = Path::new(source_file_path);
+    if source_file.exists() && source_file.is_file() {
+        let file_name = source_file
+            .file_name()
+            .map(|name| name.to_string_lossy().to_string())
+            .expect("Failed to extract file name from path");
+
+        copy(
+            PathBuf::from(source_file),
+            path_append(dest_dir_path, format!("/{}", file_name)),
+        )?;
+
+        Ok(())
+    } else {
+        Err(Error::with_context(
+            ErrorKind::FileNotFound,
+            &format!("source file {} does not exist", source_file_path),
+        ))
+    }
+}
+
+/// Helper function to create or open the tmpfs logfile
+/// The fallback log mechanism logs to a single file
+/// The filename used can be provided as an option
+pub fn open_fallback_log_file(fallback_log_filename: &str) -> Option<std::fs::File> {
+    let tmpfs_log_file = path_append("/tmp/", fallback_log_filename);
+    let log_file = match OpenOptions::new()
+        .append(true) // Append to the file, don't overwrite
+        .create(true) // Create the file if it does not exist
+        .open(tmpfs_log_file)
+    {
+        Ok(file) => Some(file),
+        Err(why) => {
+            error!(
+                "Could not open /tmp/{}, error {:?}",
+                fallback_log_filename, why
+            );
+            None
+        }
+    };
+    log_file
+}
+
+/// Helper function to persist fallback logs from tmpfs to disk
+/// This function can be called at different stages during the migration
+///
+/// # Arguments
+/// * `s2_config` - The stage2 config file
+/// * `is_new_image_flashed` - indicates if the function is being called after the
+/// the target disk has been flashed with the new os image.
+///
+/// If called before flashing, this is usually as part of error handling
+/// and we want to persist the fallback logs
+pub fn persist_fallback_log_to_data_partition(
```

Comment:
I would have used a constant instead of boolean for `is_new_imaged_flashed` parameter, but not essential to change.